analytics/src/test/java/com/segment/analytics/integrations/BasePayloadTest.kt

```diff
@@ -0,0 +1,320 @@
+            }
+
+            try {
+                builder.userId("")
+                Assert.fail()
+            } catch (e: NullPointerException) {
+                Assertions.assertThat(e).hasMessage("userId cannot be null or empty")
+            }
+        }
+    }
+
+    @Test
+    fun userId() {
+        for (builder in builders) {
+            val payload = builder.userId("user_id").build()
+            Assertions.assertThat(payload.userId()).isEqualTo("user_id")
+            Assertions.assertThat(payload).containsEntry(BasePayload.USER_ID_KEY, "user_id")
+        }
+    }
+
+    @Test
+    fun requiresUserIdOrAnonymousId() {
+        for (i in 1 until builders.size) {
+            val builder = builders[i]
+
+            try {
+                builder.build()
+                Assert.fail()
+            } catch (e: NullPointerException) {
+                Assertions.assertThat(e).hasMessage("either userId or anonymousId is required")
+            }
+        }
+    }
+
+    @Test
+    fun invalidMessageIdThrows() {
+        for (i in 1 until builders.size) {
+            val builder = builders[i]
+
+            try {
+                builder.messageId(any())
+                Assert.fail()
+            } catch (e: NullPointerException) {
+                Assertions.assertThat(e).hasMessage("messageId cannot be null or empty")
+            }
+
+            try {
+                builder.messageId("")
+                Assert.fail()
+            } catch (e: NullPointerException) {
+                Assertions.assertThat(e).hasMessage("messageId cannot be null or empty")
+            }
+        }
+    }
+
+    @Test
+    fun messageId() {
+        for (builder in builders) {
+            val payload = builder.userId("user_id").messageId("message_id").build()
+            Assertions.assertThat(payload.messageId()).isEqualTo("message_id")
+            Assertions.assertThat(payload).containsEntry(BasePayload.MESSAGE_ID, "message_id")
+        }
+    }
+
+    @Test
+    fun messageIdIsGenerated() {
+        for (builder in builders) {
+            val payload = builder.userId("user_id").build()
+            Assertions.assertThat(payload.messageId()).isNotEmpty()
+            Assertions.assertThat(payload).containsKey(BasePayload.MESSAGE_ID)
+        }
+    }
+
+    @Test
+    fun nullContextThrows() {
+        for (i in 1 until builders.size) {
+            val builder = builders[i]
+
+            try {
+                builder.context(any())
+                Assert.fail()
+            } catch (e: NullPointerException) {
+                Assertions.assertThat(e).hasMessage("context == null")
+            }
+        }
+    }
+
+    @Test
+    fun context() {
+        for (builder in builders) {
+            val payload =
+                builder.userId("user_id").context(ImmutableMap.of("foo", "bar")).build()
+            Assertions.assertThat(payload.context()).containsExactly(MapEntry.entry("foo", "bar"))
```

Comment:
is there a better alternative to `MapEntry.entry` in kotlin?

Reply:
is a `Pair` interchangable with `Entry`? If yes we should try using the `to` function for creating the pairs

Reply:
There is Map.Entry but it's an interface so it can't be directly called, I did simplify the `MapEntry.entry` to just `entry` so it would look cleaner.